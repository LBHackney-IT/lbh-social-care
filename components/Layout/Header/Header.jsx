import cx from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';

import Logo from './Logo.jsx';

const navLinks = [
  {
    name: 'Search',
    path: '/',
  },
  {
    name: 'My cases',
    path: '/cases?my_notes_only=true',
  },
  {
    name: 'Forms in progress',
    path: '/',
  },
  {
    name: 'Logout',
    path: '/logout',
  },
];

const HeaderComponent = ({ serviceName }) => {
  const { asPath } = useRouter();
  return (
    <header className="govuk-header" role="banner" data-module="govuk-header">
      <div className="govuk-header__container">
        <div className="govuk-width-container">
          <div className="govuk-header__logo">
            <Link href="/">
              <a className="govuk-header__link govuk-header__link--homepage">
                <span className="govuk-header__logotype">
                  <Logo />
                </span>
              </a>
            </Link>
          </div>
          <div className="govuk-header__content">
            <Link href="/">
              <a className="govuk-header__link govuk-header__link--service-name">
                {serviceName}
              </a>
            </Link>
            <button
              type="button"
              className="govuk-header__menu-button govuk-js-header-toggle"
              aria-controls="navigation"
              aria-label="Show or hide navigation menu"
            >
              Menu
            </button>
            <nav>
              <ul
                id="navigation"
                className="govuk-header__navigation "
                aria-label="Navigation menu"
              >
                {navLinks.map(({ name, path }) => (
                  <li
                    key={path}
                    className={cx('govuk-header__navigation-item', {
                      'govuk-header__navigation-item--active': path === asPath,
                    })}
                  >
                    <Link href={path}>
                      <a className="govuk-header__link">{name}</a>
                    </Link>
                  </li>
                ))}
              </ul>
            </nav>
          </div>
        </div>
      </div>
    </header>
  );
};

export default HeaderComponent;
