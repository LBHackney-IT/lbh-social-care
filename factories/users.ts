import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
}));

export const mockedUser = userFactory.build();
