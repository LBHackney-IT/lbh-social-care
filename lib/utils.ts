import { Form, Step, Field } from '../data/flexibleForms/forms.types';
import { Resident } from 'types';

export interface Theme {
  name: string;
  steps: Step[];
}

type InitialValue = null | string | string[] | InitialValues[];

export interface InitialValues {
  [key: string]: InitialValue;
}

/** Take a form's steps and organise them by theme for displaying in a task list */
export const groupByTheme = (form: Form): Theme[] =>
  form.steps.reduce((groups: Theme[], step) => {
    const name = step.theme;
    if (name && !groups.find((group) => group.name === name)) {
      groups.push({
        name: name,
        steps: [],
      });
    }
    const group = groups.find((group) => group.name === name);
    group?.steps.push(step);
    return groups;
  }, []);

/** Shorten a long string to a given number of words for displaying in previews */
export const truncate = (str: string, noWords: number): string => {
  if (str.split(' ').length > noWords) {
    return str.split(' ').splice(0, noWords).join(' ') + '...';
  } else {
    return str;
  }
};

const initiallyNull = new Set(['file']);
const initiallyFirstChoice = new Set(['select']);
const initiallyArray = new Set(['checkboxes', 'repeater']);

/** Generate flexible initial values for a flexible schema */
export const generateInitialValues = (
  fields: Field[],
  person?: Resident
): InitialValues => {
  const initialValues: InitialValues = {};

  fields.map((field) => {
    if (field.type === 'repeaterGroup') {
      initialValues[field.id] = [
        generateInitialValues(field.subfields || [], person),
      ];
    } else if (initiallyArray.has(field.type)) {
      initialValues[field.id] = [];
    } else if (initiallyNull.has(field.type)) {
      initialValues[field.id] = null;
    } else if (initiallyFirstChoice.has(field.type)) {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) ||
          (field.choices && field.choices[0].value)
      );
    } else {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) || ''
      );
    }
  });
  return initialValues;
};

/** Push an element into an array, without duplicates */
export const pushUnique = <T>(array: T[], newElement: T): T[] =>
  Array.from(new Set(array).add(newElement));
