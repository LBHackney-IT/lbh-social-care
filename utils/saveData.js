export const windowGlobal = typeof window !== 'undefined' && window;

export const saveData = (
  formPath,
  data,
  title,
  step,
  includesDetails,
  personDetails
) => {
  const timeStamp = new Date(Date.now()).toLocaleString().split(',')[0];
  const savedData = {
    step,
    data,
    title,
    timeStamp,
    formPath,
    includesDetails,
    ...personDetails,
  };
  try {
    windowGlobal.localStorage.setItem(formPath, JSON.stringify(savedData));
  } catch {
    deleteData(formPath);
  }
};

export const getData = (formPath) => {
  try {
    return JSON.parse(windowGlobal.localStorage.getItem(formPath));
  } catch {
    deleteData(formPath);
  }
};

export const getDataIncludes = (includes) => {
  if (windowGlobal.localStorage) {
    try {
      const data = Object.fromEntries(
        Object.entries(windowGlobal.localStorage)
          .filter(([key]) => key.includes(includes))
          .map(([key, value]) => [key, JSON.parse(value)])
      );
      return Object.keys(data).length > 0 ? data : null;
    } catch {
      windowGlobal.localStorage.clear();
      return null;
    }
  }
};

export const deleteData = (formPath) => {
  windowGlobal.localStorage.removeItem(formPath);
};
