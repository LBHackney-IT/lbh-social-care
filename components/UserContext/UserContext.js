import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
