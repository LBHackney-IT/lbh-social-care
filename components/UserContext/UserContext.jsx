import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
