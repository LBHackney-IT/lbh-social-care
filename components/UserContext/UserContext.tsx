import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
