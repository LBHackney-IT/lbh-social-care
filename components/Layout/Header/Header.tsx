import React, { useState, useEffect } from 'react';
import cx from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';

import { useAuth } from 'components/UserContext/UserContext';
import { getData } from 'utils/saveData';
import { getUserType } from 'utils/user';
import Logo from './Logo';

const loggedNavLinks = [
  {
    name: 'Search',
    path: '/',
    isSelected: ({ asPath, pathname }: { asPath: string; pathname: string }) =>
      pathname === '/' ||
      (pathname === '/cases' && asPath !== '/cases?my_notes_only=true'),
  },
  {
    name: 'My records',
    path: '/cases?my_notes_only=true',
    isSelected: ({ asPath }: { asPath: string }) =>
      asPath === '/cases?my_notes_only=true',
  },
  {
    name: 'Forms in progress',
    path: '/form-in-progress',
  },
  {
    name: 'Manage workers',
    path: '/workers',
  },
  {
    name: 'Logout',
    path: '/logout',
  },
];

const HeaderComponent = ({
  serviceName,
}: {
  serviceName: string;
}): React.ReactElement => {
  const { user } = useAuth();
  const { pathname, asPath } = useRouter();
  const [navLinks, setNavLinks] = useState<typeof loggedNavLinks>();
  useEffect(() => {
    if (user) {
      const savedForms = getData();
      setNavLinks(
        loggedNavLinks
          .filter(({ name }) => name !== 'Forms in progress' || savedForms)
          .filter(
            ({ name }) => name !== 'Manage workers' || user.hasDevPermissions
          )
      );
    }
  }, [user, pathname]);
  return (
    <header className="lbh-header ">
      <div className="lbh-header__main">
        <div className="lbh-container lbh-header__wrapper">
          <div className="lbh-header__title">
            <a href="/" className="lbh-header__title-link">
              <Logo />
              <span className="lbh-header__logo-text"> Hackney </span>
              <span className="lbh-header__service-name">Social care</span>
              <span className="govuk-tag lbh-tag">
                {user && getUserType(user)}
              </span>
            </a>
          </div>
          <nav className="lbh-header__links" aria-label="Navigation menu">
            {navLinks && (
              <>
                {navLinks.map(({ name, path, isSelected }) => (
                  <Link href={path} key={path}>
                    <a className="govuk-header__link">{name}</a>
                  </Link>
                ))}
              </>
            )}
          </nav>
        </div>
      </div>
    </header>
  );
};

export default HeaderComponent;
