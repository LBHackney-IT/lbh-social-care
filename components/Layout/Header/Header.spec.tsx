import { render } from '@testing-library/react';

import Header from './Header';
import { UserContext } from 'components/UserContext/UserContext';
import { mockedUser } from 'factories/users';
import * as saveData from 'utils/saveData';

let mockedUseRouter = { pathname: 'pathname' };

jest.mock('next/router', () => ({
  asPath: 'path',
  useRouter: () => mockedUseRouter,
}));

jest.mock('./Logo', () => () => 'MockedLogo');

describe('Header component', () => {
  const props = {
    serviceName: 'Foo',
  };
  it('should render header links with form in progress link', () => {
    jest.spyOn(saveData, 'getData').mockImplementationOnce(() => ({
      first: {
        data: { id: '123' },
        formPath: '/form/foo-bar/',
        timestamp: '22/12/2020',
        title: 'Foo Bar',
        step: 'foo',
        deleteForm: jest.fn(),
      },
    }));
    const { getByText } = render(
      <UserContext.Provider
        value={{
          user: mockedUser,
        }}
      >
        <Header {...props} />
      </UserContext.Provider>
    );

    expect(getByText('Search')).toBeInTheDocument();
    expect(getByText('Sign out')).toBeInTheDocument();
  });

  it('should render service name but no header links', () => {
    const { getByText, queryByText } = render(
      <UserContext.Provider
        value={{
          user: undefined,
        }}
      >
        <Header {...props} />
      </UserContext.Provider>
    );
    expect(getByText('Foo')).toBeInTheDocument();
    expect(queryByText('Search')).not.toBeInTheDocument();
    expect(queryByText('Sign out')).not.toBeInTheDocument();
  });

  it('should set active the correct link', () => {
    mockedUseRouter = {
      pathname: '/my-records',
    };
    const { getByText, asFragment } = render(
      <UserContext.Provider
        value={{
          user: mockedUser,
        }}
      >
        <Header {...props} />
      </UserContext.Provider>
    );
    expect(getByText('My work space')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
