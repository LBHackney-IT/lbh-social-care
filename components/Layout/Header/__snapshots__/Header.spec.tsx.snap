// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should set active the correct link 1`] = `
<DocumentFragment>
  <header
    class="lbh-header "
  >
    <div
      class="lbh-header__main"
    >
      <div
        class="lbh-container lbh-header__wrapper"
      >
        <div
          class="lbh-header__title"
        >
          <a
            class="lbh-header__title-link"
            href="/"
          >
            MockedLogo
            <span
              class="lbh-header__logo-text"
            >
               Hackney 
            </span>
            <span
              class="lbh-header__service-name"
            >
              Foo
            </span>
            <span
              class="govuk-tag lbh-tag lbh-tag--green"
            >
              Admin
            </span>
          </a>
        </div>
        <nav
          aria-label="Navigation menu"
          class="lbh-header__links"
        >
          <a
            class="govuk-header__link"
            href="/search"
          >
            Search
          </a>
          <a
            class="govuk-header__link"
            href="/"
          >
            My work space
          </a>
          <a
            class="govuk-header__link"
            href="/logout"
          >
            Sign out
          </a>
        </nav>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
